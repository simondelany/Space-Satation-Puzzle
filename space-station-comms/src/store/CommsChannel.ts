export class CommsChannel {
  private listeners: ((msg: string) => void)[] = []

  public broadcast (msg: string) {
    for (const msgHandler of this.listeners) {
      msgHandler(msg)
    }
  }

  public subscribe (msgHandler: ((msg: string) => void)) {
    this.listeners.push(msgHandler)
  }

  public unsubscribe (msgHandler: ((msg: string) => void)) {
    this.listeners = this.listeners.filter((handler: ((msg: string) => void)) => {
      return handler !== msgHandler
    })
  }
}
