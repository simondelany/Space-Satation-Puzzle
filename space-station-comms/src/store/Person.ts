interface CareerPosition {
    title: string;
}

export interface Person {
    uuid: string,
    status: string,
    title: string,
    firstName: string,
    lastName: string,
    transferTime: number,
    location: number,
    position: CareerPosition,
    selected?: boolean
}
