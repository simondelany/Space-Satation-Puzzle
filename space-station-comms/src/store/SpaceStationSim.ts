export class SpaceStationSim {
  private broadcast: (msg: string) => void
  public ready: boolean

  constructor (broadcaster: (msg: string) => void) {
    // We'll use this to send out messages to the user
    this.broadcast = broadcaster
    this.ready = false
  }

  private startScript = [
    3,
    'The Space Station has split in two!',
    2,
    'We\'re not sure how this happened...',
    1,
    '...the module we\'re all stranded on is rapidly loosing oxygen!',
    3,
    'I can see another section of the ISS is still intact!',
    2,
    'We need to figure out how to get us all off the damaged module.',
    1,
    'We only have 21 minutes before the oxygen runs out...',
    2,
    '...I\'ve looked everywhere but it seems only 2 spacesuits have survived intact...',
    1,
    'You\'ve got to help us figure this out before the time runs out!',
    5
  ]

  public startSimulation (callback: () => void) {
    this.broadcastScript(this.startScript, callback)
  }

  private broadcastScript (script: (string | number)[], callback: () => void) {
    console.dir(script)
    this.handleNextLine(0, callback)
  }

  private handleNextLine = (nextIndex: number, callback: () => void) => {
    if (nextIndex < this.startScript.length) {
      const line: string | number = this.startScript[nextIndex]
      switch (typeof line) {
        case 'string':
          this.broadcast(line)
          console.log(line)
          this.handleNextLine(nextIndex + 1, callback)
          break
        case 'number':
          setTimeout(this.handleNextLine, line * 1200, nextIndex + 1, callback)
          break
      }
    } else {
      callback()
    }
  }
}
