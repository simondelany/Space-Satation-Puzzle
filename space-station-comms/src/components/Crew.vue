<template>
    <div class="crew">
        <div class="crew__block"
            v-bind:class="{selected: data.selected}"
            draggable="true"
            @click="handleClick">
            <p class="person-name">{{ data.firstName }}</p>
        </div>
    </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import { Person } from '@/store/Person.ts'

@Component
export default class Crew extends Vue {
  @Prop() private data!: Person

  private handleClick () {
    if (this.data.selected) {
      this.$emit('deselectPerson', this.data)
    } else {
      this.$emit('selectPerson', this.data)
    }
  }
}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
@import "@/styles/definitions/_variables.scss";
  .crew {
    padding: 0.5vmin;

    &__block {
      background-color: rgb(36, 33, 65);
      height: 100%;
      padding: 0;
      margin-top: auto;
    }

    &__block.selected {
      background-color: blue;
    }
  }

</style>
