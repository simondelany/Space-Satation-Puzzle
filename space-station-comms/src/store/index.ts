import Vue from 'vue'
import Vuex from 'vuex'
import Axios from 'axios'
import { Person } from '@/store/Person'
import { CommsChannel } from '@/store/CommsChannel'
import { SpaceStationSim } from '@/store/SpaceStationSim'

Vue.use(Vuex)

const crew: Person[] = []
const commsChannel = new CommsChannel()
const broadcaster = (msg: string) => {
  commsChannel.broadcast(msg)
}
const sim = new SpaceStationSim(broadcaster)

export default new Vuex.Store({
  state: {
    ready: false,
    crew,
    suitPosition: 0,
    remainingTime: 21
  },
  mutations: {
    loadCrewData (state, data) {
      state.crew = data
    },
    movePeople (state, data) {
      // move the crew
      state.crew = state.crew.map((c: Person) => {
        const beingMoved = data.people.find((p: Person) => {
          return p.uuid === c.uuid
        })
        if (beingMoved) {
          c.location = data.position
        }
        return c
      })

      // When completing a transit
      if (data.position !== 2) {
        // switch suit position
        state.suitPosition = 1 - state.suitPosition

        // clear the selections
        state.crew = state.crew.map((c: Person) => {
          c.selected = false
          return c
        })

        state.remainingTime -= data.transferTime
      }
    },
    updateSelectedCrew (state, selected: Person[]) {
      state.crew = state.crew.map((c: Person) => {
        const beingSelected = selected.find((s: Person) => {
          return s.uuid === c.uuid
        })
        if (beingSelected) {
          c.selected = true
        } else {
          c.selected = false
        }
        return c
      })
    },
    openCommsChannel (state, msgHandler: ((msg: string) => void)) {
      commsChannel.subscribe(msgHandler)
    },
    ready (state) {
      state.ready = true
    }
  },
  actions: {
    fetchCrewData (store) {
      // start the simulation - intro messages will show while we load the data
      sim.startSimulation(() => {
        store.commit('ready')
      })

      Axios.get(process.env.VUE_APP_API_EMPLOYEES).then(employeeRequest => {
        let employees: Person[] = []
        if (employeeRequest.status === 200) {
          employees = employeeRequest.data.map((person: Person) => {
            person.location = 0
            person.selected = false
            person.status = 'OK'
            return person
          })
        }
        store.commit('loadCrewData', employees)
      }).catch(() => {
        console.warn('Could not load employee physical assesments!')
      })
    },
    startTransfer: (store) => {
      const selected = store.getters.getSelectedCrew()
      const transferTime = Math.max(...selected.map((s: Person) => s.transferTime))
      commsChannel.broadcast('Commencing crew transfer...')
      commsChannel.broadcast(`E.T.A. to completion: ${transferTime} minute${(transferTime > 1) ? 's' : ''}`)

      // move selected into the void of space
      store.commit(
        'movePeople',
        {
          people: selected,
          position: 2,
          transferTime
        }
      )

      setTimeout(() => {
        // complete transfer if we have enough time remaining
        store.commit(
          'movePeople',
          {
            people: (store.state.remainingTime >= transferTime) ? selected : [],
            position: (store.state.remainingTime >= transferTime) ? 1 - store.state.suitPosition : store.state.suitPosition,
            transferTime
          }
        )

        const stillInTransit = store.getters.getCrewInTransit()

        if (!stillInTransit.length) {
          commsChannel.broadcast('Crew transfer complete.')
          commsChannel.broadcast('They made it across safely!')
        }

        if (store.state.remainingTime > 0) {
          commsChannel.broadcast(`Only ${store.state.remainingTime} minute${(store.state.remainingTime > 1) ? 's' : ''} left!`)
        } else {
          const lost = [].concat(store.getters.getCrewA(), stillInTransit)
          if (lost.length) {
            store.commit('loadCrewData', store.getters.getCrewB())
            commsChannel.broadcast(`... automated critical message ...\nModule Alpha destroyed!\nCrew Lost:\n* ${lost.map((c: Person) => c.firstName).join('\n* ')}`)
          } else {
            commsChannel.broadcast('You did it! We\'re all safely across!')
          }
        }
      }, transferTime * 1000)
    }
  },
  getters: {
    getCrewA: (state) => () => {
      return state.crew.filter((person: Person) => {
        return person.location === 0
      })
    },
    getCrewB: (state) => () => {
      return state.crew.filter((person: Person) => {
        return person.location === 1
      })
    },
    getCrewInTransit: (state) => () => {
      return state.crew.filter((person: Person) => {
        return person.location === 2
      })
    },
    getSelectedCrew: (state) => () => {
      return state.crew.filter((person: Person) => {
        return (person.selected === true) && (person.location !== 2)
      })
    },
    isReady: (state) => () => {
      return state.ready
    }
  }
})
