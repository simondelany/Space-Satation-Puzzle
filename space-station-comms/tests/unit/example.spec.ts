import { shallowMount } from '@vue/test-utils'
import Crew from '@/components/Crew.vue'
import { Person } from '@/store/Person'

describe('Crew.vue', () => {
  it('renders props.data when passed', () => {
    const data: Person = {
      uuid: 'd00f2fef-75e3-44a2-bdde-87acf04e6331',
      firstName: 'Neil',
      lastName: 'Fitzpatrick',
      title: 'Mr.',
      position: {
        title: 'Pilot'
      },
      location: 0,
      status: 'OK',
      transferTime: 1
    }
    const wrapper = shallowMount(Crew, {
      propsData: { data }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
