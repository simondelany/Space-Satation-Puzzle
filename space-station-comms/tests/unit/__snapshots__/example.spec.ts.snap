// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crew.vue renders props.data when passed 1`] = `
<div class="crew">
  <div draggable="true" class="crew__block">
    <p class="person-name">Neil</p>
  </div>
</div>
`;
